src/linear_algebra/dual.lean

```diff
@@ -0,0 +1,244 @@
+    contradiction }
+end
+
+theorem to_dual_ker : h.to_dual.ker = ⊥ :=
+ker_eq_bot'.mpr h.to_dual_inj
+
+theorem to_dual_range [fin : fintype B] : h.to_dual.range = ⊤ :=
+begin
+  rw eq_top_iff',
+  intro f,
+  rw linear_map.mem_range,
+  let lin_comb : B →₀ K := finsupp.on_finset fin.elems (λ b, f.to_fun b) _,
+  let emb := embedding.subtype B,
+  { use lc.total K V (finsupp.emb_domain emb lin_comb),
+    apply h.ext,
+    intros x hx,
+    rw [h.to_dual_eq_repr _ x hx, repr_total _],
+    have emb_x : x = emb ⟨x, hx⟩, from rfl,
+    { rw [emb_x, finsupp.emb_domain_apply emb lin_comb _, ← emb_x], simpa },
+    { rw [lc.mem_supported, finsupp.support_emb_domain, finset.map_eq_image, finset.coe_image],
+      apply subtype.val_image_subset } },
+  { intros a _,
+    apply fin.complete }
+end
+
+def dual_basis : set (dual K V) := h.to_dual '' B
+
+theorem dual_lin_independent : linear_independent K h.dual_basis :=
+begin
+  apply linear_independent.image h.1,
+  rw to_dual_ker,
+  exact disjoint_bot_right
+end
+
+def to_dual_equiv [fintype B] : V ≃ₗ[K] (dual K V) :=
+linear_equiv.of_bijective h.to_dual h.to_dual_ker h.to_dual_range
+
+theorem dual_basis_is_basis [fintype B] : is_basis K h.dual_basis :=
+h.to_dual_equiv.is_basis h
+
+@[simp] lemma to_dual_to_dual [fintype B] :
+  (h.dual_basis_is_basis.to_dual).comp h.to_dual = eval K V :=
+begin
+  apply h.ext,
+  intros b hb,
+  apply h.dual_basis_is_basis.ext,
+  intros c hc,
+  dunfold eval,
+  rw [linear_map.flip_apply, linear_map.id_apply, linear_map.comp_apply,
+      to_dual_apply h.dual_basis_is_basis (h.to_dual b) _ c hc],
+  { dsimp [dual_basis] at hc,
+    rw set.mem_image at hc,
+    rcases hc with ⟨b', hb', rfl⟩,
+    rw to_dual_apply,
+    split_ifs with h₁ h₂; try {refl},
+    { have hh : b = b', from (ker_eq_bot.1 h.to_dual_ker h₁).symm, contradiction },
+    { subst h_1, contradiction },
+    assumption' },
+  { dunfold dual_basis,
+    exact set.mem_image_of_mem h.to_dual hb }
+end
+
+theorem dual_dim_eq [fintype B] :
+  cardinal.lift.{v (max u v)} (dim K V) = dim K (dual K V) :=
+by { rw linear_equiv.dim_eq_lift h.to_dual_equiv, apply cardinal.lift_id }
+
+end is_basis
+
+
+namespace vector_space
+
+universes u v
+variables {K : Type u} {V : Type v}
+variables [discrete_field K] [add_comm_group V] [vector_space K V]
+open module module.dual submodule linear_map cardinal is_basis
+
+theorem eval_ker : (eval K V).ker = ⊥ :=
+begin
+  rw ker_eq_bot',
+  intros v h,
+  rw linear_map.ext_iff at h,
+  by_contradiction H,
+  rcases exists_subset_is_basis (linear_independent_singleton H) with ⟨b, hv, hb⟩,
+  swap 4, assumption,
+  have hv' : v ∈ b := hv (set.mem_singleton v),
+  let hx := h (hb.to_dual v),
+  erw [eval_apply, to_dual_apply _ _ hv' _ hv', if_pos rfl, zero_apply _] at hx,
+  exact one_ne_zero hx
+end
+
+lemma exists_is_basis_fintype (h : dim K V < omega) :
+  ∃B : (set V), (is_basis K B) ∧ nonempty (fintype B) :=
```

Comment:
I don't know what is best here. Should we use `nonempty (fintype B)` or `finite B`?